server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java

```diff
@@ -0,0 +1,488 @@
+            @Override
+            public void onResponse(BaseNodesResponse<T> response) {
+                processAsyncFetch(response.getNodes(), response.failures(), fetchingRound);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                List<FailedNodeException> failures = new ArrayList<>(nodes.length);
+                for (final DiscoveryNode node : nodes) {
+                    failures.add(new FailedNodeException(node.getId(), "total failure in fetching", e));
+                }
+                processAsyncFetch(null, failures, fetchingRound);
+            }
+        });
+    }
+
+    /**
+     * The result of a fetch operation. Make sure to first check {@link #hasData()} before
+     * fetching the actual data.
+     */
+    public static class FetchResult<T extends BaseNodeResponse> {
+
+        private final Map<DiscoveryNode, T> data;
+        private Map<ShardId, Set<String>> ignoredShardToNodes = new HashMap<>();
+
+        public FetchResult(Map<DiscoveryNode, T> data, Map<ShardId, Set<String>> ignoreNodes) {
+            this.data = data;
+            this.ignoredShardToNodes = ignoreNodes;
+        }
+
+        /**
+         * Does the result actually contain data? If not, then there are on going fetch
+         * operations happening, and it should wait for it.
+         */
+        public boolean hasData() {
+            return data != null;
+        }
+
+        /**
+         * Returns the actual data, note, make sure to check {@link #hasData()} first and
+         * only use this when there is an actual data.
+         */
+        public Map<DiscoveryNode, T> getData() {
+            assert data != null : "getData should only be called if there is data to be fetched, please check hasData first";
+            return this.data;
+        }
+
+        /**
+         * Process any changes needed to the allocation based on this fetch result.
+         */
+        public void processAllocation(RoutingAllocation allocation) {
+            for(Map.Entry<ShardId, Set<String>> entry : ignoredShardToNodes.entrySet()) {
+                ShardId shardId = entry.getKey();
+                Set<String> ignoreNodes = entry.getValue();
+                if (ignoreNodes.isEmpty() == false) {
+                    ignoreNodes.forEach(nodeId -> allocation.addIgnoreShardForNode(shardId, nodeId));
+                }
+            }
+
+        }
+    }
+
+    /**
+     * The result of a fetch operation. Make sure to first check {@link #hasData()} before
+     * fetching the actual data.
+     */
+    public static class AdaptedResultsForShard<T extends BaseNodeResponse> {
+
+        private final ShardId shardId;
+        private final Map<DiscoveryNode, T> data;
+        private final Set<String> ignoreNodes;
+
+        public AdaptedResultsForShard(ShardId shardId, Map<DiscoveryNode, T> data, Set<String> ignoreNodes) {
```

Comment:
Better name **ShardFetchResult**

Reply:
This is used as Adapter, and just kept as placeholder, we will vary its implementation in next Sub PR for assignment. We will see the naming according to the usage

Reply:
We'll not use adapters, so you can make changes accordingly.